File: experiments/control-run/go/contract/contract.go
// Package contract checks requires, ensures, and invariants. A failed
// check returns an error that carries the condition's text; it never panics.
package contract

// Violation is a failed contract: which kind, and the condition as written.
type Violation struct {
	Kind string
	Cond string
}

func (v *Violation) Error() string { return v.Kind + " failed: " + v.Cond }

func check(kind string, ok bool, cond string) error {
	if ok {
		return nil
	}
	return &Violation{Kind: kind, Cond: cond}
}

// Require checks a precondition.
func Require(ok bool, cond string) error { return check("requires", ok, cond) }

// Ensure checks a postcondition.
func Ensure(ok bool, cond string) error { return check("ensures", ok, cond) }

// Invariant checks a property that holds after every operation.
func Invariant(ok bool, cond string) error { return check("invariant", ok, cond) }

// Never fails when the thing that must never happen did.
func Never(happened bool, cond string) error { return check("never", !happened, cond) }

File: experiments/control-run/python/logstat/src/logstat/contract.py
"""Contracts as explicit checks. A failure names the condition that did not hold.

These are not `assert` statements, so `python -O` does not remove them.
"""


class ContractError(Exception):
    """A requires, ensures, or invariant check failed: a bug in logstat, not bad input."""


def require(condition: bool, text: str) -> None:
    """A precondition the caller must meet."""
    if not condition:
        raise ContractError(f"requires {text}")


def ensure(condition: bool, text: str) -> None:
    """A postcondition the function promises."""
    if not condition:
        raise ContractError(f"ensures {text}")


def invariant(condition: bool, text: str) -> None:
    """A condition that holds after every change of state."""
    if not condition:
        raise ContractError(f"invariant {text}")
